packages/expression-evaluator/lib/evaluators/ExpressionEvaluator.ts

```diff
@@ -1,195 +1,21 @@
-import type { IActionContext, IExpressionEvaluator } from '@comunica/types';
+import type { IExpressionEvaluator } from '@comunica/types';
 import type * as RDF from '@rdfjs/types';
-import { LRUCache } from 'lru-cache';
-import type { Algebra as Alg } from 'sparqlalgebrajs';
 import type * as E from '../expressions';
-import { regularFunctions } from '../functions';
-import type { FunctionArgumentsCache } from '../functions/OverloadTree';
-import { AlgebraTransformer } from '../transformers/AlgebraTransformer';
-import { TermTransformer } from '../transformers/TermTransformer';
-import * as C from '../util/Consts';
-import type { ITimeZoneRepresentation } from '../util/DateTimeHelpers';
-import { extractTimeZone } from '../util/DateTimeHelpers';
-import * as Err from '../util/Errors';
-import type { SuperTypeCallback, TypeCache } from '../util/TypeHandling';
-import type { ICompleteEEContext } from './evaluatorHelpers/AsyncRecursiveEvaluator';
-import { AsyncRecursiveEvaluator } from './evaluatorHelpers/AsyncRecursiveEvaluator';
-
-export type AsyncExtensionFunction = (args: RDF.Term[]) => Promise<RDF.Term>;
-export type AsyncExtensionFunctionCreator = (functionNamedNode: RDF.NamedNode) => AsyncExtensionFunction | undefined;
-
-// TODO: make this fields of the EE
-export interface IAsyncEvaluatorContext {
-  exists: (expression: Alg.ExistenceExpression, mapping: RDF.Bindings) => Promise<boolean>;
-  aggregate?: (expression: Alg.AggregateExpression) => Promise<RDF.Term>;
-  bnode: (input?: string) => Promise<RDF.BlankNode>;
-  extensionFunctionCreator?: AsyncExtensionFunctionCreator;
-  now?: Date;
-  baseIRI?: string;
-  typeCache?: TypeCache;
-  getSuperType?: SuperTypeCallback;
-  functionArgumentsCache?: FunctionArgumentsCache;
-  defaultTimeZone?: ITimeZoneRepresentation;
-  actionContext: IActionContext;
-}
+import type { ContextualizedEvaluator } from './ContextualizedEvaluator';
 
 export class ExpressionEvaluator implements IExpressionEvaluator {
```

Comment:
This class doesn't seem to be doing much on its own. Is there a reason why `ContextualizedEvaluator` can't just be merged into this one?

Reply:
Yes! So our translation function is now async (translating means we find the correct function definition for an operator, which means we need a mediator call. A constructor can not be async, so I decided to solve that issue this way. Another way could be using an async `initialize` function.

Reply:
> So our translation function is now async (translating means we find the correct function definition for an operator, which means we need a mediator call. A constructor can not be async

I don't understand the relationship between this ^ and the `ContextualizedEvaluator`.

Just blindly moving everything from inside of the `ContextualizedEvaluator` towards this class wouldn't change anything, right?

Reply:
Hmm, so before evaluation, you need to translate the `Algebra.Expressions` to `Sparqlee.Expressions`, this means that the operators should be translated. You do not want to translate each time you evaluate. Comunica wants to evaluate the same Algebra but with different bindings. What I try to do here is create an easy API that translates the Algebra within the constructor. (Which wouldn't be possible when just merging the files because constructors can not be async)

Reply:
Ok, I understand the need now.

So the main purpose of `ContextualizedEvaluator` is to translate between the two expression types, which is an async operation.
Would it make sense to call it something like `ExpressionTranslator`?

I'm not too sure yet if the following methods really make sense to be in that class:

- `evaluateAsInternal`
- `evalFunction`
- `evalExistence`
- `evalAggregate`

Should they be moved into `ExpressionEvaluator`? Or do they need to be moved to a new class?

Reply:
> So the main purpose of ContextualizedEvaluator is to translate between the two expression types, which is an async operation.

Euhm, no :sweat_smile: . `ContextualizedEvaluator` is more of an `InternalEvaluator` it is able to evaluate something that has already been translated.

Reply:
@rubensworks I now renamed it to `InternalEvaluator` and extracted the config in ActionContextKeys. From what I can see, that's much prettier. If you agree you can resolve this conversation :)

Reply:
Looks good!